File: frontend/src/lib/account-client/sessions.ts
import { API_BASE } from "@/lib/session";
import { authHeaders, parseOrThrow } from "@/lib/api-client/shared";
import { setToken } from "@/lib/auth";

/** Revokes every other device's access token via a token_version bump,
 * then swaps in the fresh token the backend issues for THIS device so
 * calling it doesn't also sign the caller out. */
export async function signOutOtherDevices(): Promise<void> {
  const res = await fetch(`${API_BASE}/api/account/sessions/sign-out-others`, {
    method: "POST",
    headers: authHeaders(),
  });
  const data = await parseOrThrow<{ access_token: string }>(res);
  setToken(data.access_token);
}

File: Backend/app/routers/account/sessions.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session

from app.database import get_db
from app.models.user import User
from app.auth.dependencies import get_current_user
from app.services.account import sign_out_other_devices

router = APIRouter()


@router.post("/api/account/sessions/sign-out-others")
def sign_out_others(user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    """Real revocation, not decorative: every access token issued before
    this call fails get_current_user's token_version check on its next
    request. Returns a fresh token for the device that asked, so calling
    this doesn't also sign the caller out."""
    new_token = sign_out_other_devices(db, user)
    return {"access_token": new_token, "token_type": "bearer"}
